### ISSUE TITLE
Should distortion admit a distance parameter

### ISSUE THREAD

**Issue by FreddieWitherden** [lensfun/lensfun#1292]
I am wondering if the distortion correction element should incorporate support for a distance parameter, much like what the vignetting element supports.  This is useful for lenses which suffer from substantial focus breathing where the amount of distortion at close focus distances can be very different from the amount of distortion at infinity.

Background: I am reverse engineering the distortion correction parameters which Sony cameras embed into their RAW files and observe that for some prime lenses the embedded parameters can be quite different depending on the distance at which the lens was focused.

For example, for the Sony FE 50mm F1.8 at a 0.62m distance the 16 distortion coefficients read: `1 0 -2 -4 -8 -13 -18 -25 -33 -42 -52 -64 -75 -89 -103 -119` whereas at 34.72m the coefficients read: `1 0 -1 -3 -4 -5 -7 -10 -13 -16 -19 -23 -27 -31 -33 -36`.  This corresponds to a factor of three difference! (And yes, the coefficients can be interpreted in a linear fashion given that they appear to define the knots of a spline.)

**Comment by bronger** [comment]
First of all: Focus breathing is a significant effect, and I myself have created two distinct profiles for one of my lenses, one at 50cm and one at infinity, because I use this lens for digitising documents and need a perfect profile at 50cm.

That said, I think there are also other things to consider here besides what’s most accurate.  Interpolating is an issue: So far, only vignetting profiles depend on more than one value, and its interpolation algorithm is sub-optimal. A better algorithm would be much more complicated, however.

Moreover, it may make making good profiles more difficult. Even now, many people avoid taking pictures at large distances and send close-up images of grid on paper etc.  Thus, it may lead to more confusion, which already is a serious problem when processing sample images and communicating with their originators.

Still, I would support such an extension if someone has a solution for the interpolation issue.

**Comment by FreddieWitherden** [comment]
I agree that interpolation is an issue and not one which is easy to solve.

However, I am not sure that I totally agree that it will make creating good profiles any more difficult.  The reality at the moment is that it is very difficult to create a good profile, and due to this lensfun contains a lot of 'okay' profiles — substantially better than nothing but far from ideal.  All this would do is help make people more away that a profile may not be 'right' for them (say because it was taking at 0.5m whilst their picture is focused at infinity).

This is why I am interesting in working with the manufacturer provided coefficients and using them to fit a profile.  In this regime data is very easy to come by and this one can easily do a very fine sampling of a lens just by racking its focus and taking ~100 shots.  This in turn puts less stress on the interpolation algorithm.

**Comment by blueglyph** [comment]
I had the same question and just saw here that it was raised 4 years ago, then abandoned.

I noticed a slight difference between correction parameters calculated from shots at about 8 m with a 35 mm, so near infinity for that lens, and shots taken at 30 cm, so near the lens's minimum focus distance. The lens has a very small pincushion distortion in both cases.

If I apply the correction from the 1st set, it's fine on shots of the first set, but there's still some pincushion on the 2nd set, easy to spot by drawing straight lines on the final image. I'd say the distortion was about 2-3 pixels at worst. So the distortion seems a little greater when the focus is near.

It's admittedly not big, but it was small to begin with, so I expect other lenses (like macros) could display a more important dependency to the distance. It would really be worth adding an optional distance parameter, but I don't know how to do it without breaking the compatibility in programs that are already using the database (though it's not reached 1.0 yet, so changes should be expected). 

Would an extra attribute be fine? Maybe someone already looked at it in programs like Darktable and Rawtherapee? I did a quick test with Affinity Photo 2 (2.6) and it didn't seem to bother the tool - see the example below.

```
<distortion model="ptlens" focal="35" a="0.00726" b="-0.01914" c="0.0101" distance="0.3" />
<distortion model="ptlens" focal="35" a="0.00726" b="-0.01914" c="0.0101" distance="1000" />
```

What I wonder, though, is how the distance is known by the tool when it needs to choose the correction? It's not in the EXIF metadata, so I assume it'll have to be entered manually by the users who prefer a higher precision in the final result.

**Comment by FreddieWitherden** [comment]
Focus distance is usually available in EXIF data.  It is something the camera usually needs to know in order for TTL flash metering to work effectively and hence is typically written out to metadata by rote.

**Comment by blueglyph** [comment]
> Focus distance is usually available in EXIF data. It is something the camera usually needs to know in order for TTL flash metering to work effectively and hence is typically written out to metadata by rote.

I can't seem to find that information; what's the name? The only information I saw was `SubjectDistanceRange - Unknown`, but it's "unknown" for all the lenses I've checked on a Nikon Z6III.

**Comment by FreddieWitherden** [comment]
> > Focus distance is usually available in EXIF data. It is something the camera usually needs to know in order for TTL flash metering to work effectively and hence is typically written out to metadata by rote.
> 
> I can't seem to find that information; what's the name? The only information I saw was `SubjectDistanceRange - Unknown`, but it's "unknown" for all the lenses I've checked on a Nikon Z6III.

In my NEF files (taken on a D5300 some years ago) the relevant field is "Focus Distance" in the "NikonLd3" Makernote block.

**Comment by blueglyph** [comment]
> > > Focus distance is usually available in EXIF data. It is something the camera usually needs to know in order for TTL flash metering to work effectively and hence is typically written out to metadata by rote.
> > 
> > 
> > I can't seem to find that information; what's the name? The only information I saw was `SubjectDistanceRange - Unknown`, but it's "unknown" for all the lenses I've checked on a Nikon Z6III.
> 
> In my NEF files (taken on a D5300 some years ago) the relevant field is "Focus Distance" in the "NikonLd3" Makernote block.

I never had that information in the EXIF, neither with my Z6III nor with a Pentax K5. It must be limited to some models or to flash shots, so we can't rely on that information for everyone, but it could be used when present.
